esphome/core/bytebuffer.cpp

```diff
@@ -1,19 +1,70 @@
 #include "bytebuffer.h"
 #include <cassert>
+#include <cstring>
 
 namespace esphome {
 
-ByteBuffer ByteBuffer::create(size_t capacity) {
+ByteBuffer ByteBuffer::create(size_t capacity, Endian endianness) {
   std::vector<uint8_t> data(capacity);
-  return {data};
+  ByteBuffer buffer = {data};
+  buffer.endianness_ = endianness;
+  return buffer;
 }
 
-ByteBuffer ByteBuffer::wrap(uint8_t *ptr, size_t len) {
+ByteBuffer ByteBuffer::wrap(const uint8_t *ptr, size_t len, Endian endianness) {
   std::vector<uint8_t> data(ptr, ptr + len);
-  return {data};
+  ByteBuffer buffer = {data};
+  buffer.endianness_ = endianness;
+  return buffer;
 }
 
-ByteBuffer ByteBuffer::wrap(std::vector<uint8_t> data) { return {std::move(data)}; }
+ByteBuffer ByteBuffer::wrap(std::vector<uint8_t> data, Endian endianness) {
+  ByteBuffer buffer = {std::move(data)};
```

Comment:
```suggestion
ByteBuffer ByteBuffer::wrap(std::vector<uint8_t> const &data, Endian endianness) {
  ByteBuffer buffer = {data};
```

Constructing a new vector always does a copy. But I've realised there are unnecessary copies in the existing constructor and create method that should be streamlined (I was not thinking about move/copy semantics at the time.) It will be simpler if I just push some changes to your branch - are you ok with that?

Reply:
@clydebarrow, of course! I think I gave permissions when opening the PR.

Reply:
Yes, I have the power :-) But always like to ask first. I'll get onto that tomorrow, it's late here. I think we're nearly there with this though.